src/neural/cuda/network_cuda.cc

```diff
@@ -280,10 +283,9 @@ class CudaNetwork : public Network {
     scratch_size_ = std::max(scratch_size_, 2 * transformed_tensor_size);
 
     // Attention policy head may need more memory
-    // (We also split the allocations into two parts, so need 2x)
     const size_t attentionSize =
         getMaxAttentionHeadSize(weights, max_batch_size_);
-    scratch_size_ = std::max(scratch_size_, 2 * attentionSize);
+    scratch_size_ = std::max(scratch_size_, attentionSize);
```

Comment:
hrmm not sure about this. Allocations are still split?

Reply:
I added "size * 3" in getMaxAttentionHeadSize() which should take care of this part (It's not very obvious, but the code doesn't actually use second part of the allocation when using them for batched gemms)

Reply:
If getMaxAttentionHeadSize uses multipliers for each phase as appropriate - depending on whether they use the bottom half only or top as well as well as the actual size after batching - then yes it makes sense to drop the 2 here.

Reply:
I have simplified intermediate memory allocations a bit (also reduced from 5 scratch allocations to 4). Hopefully it's accurate now.